src/main/java/VASSAL/configure/ConfigureTree.java

```diff
@@ -284,6 +321,205 @@ protected JPopupMenu buildPopupMenu(final Configurable target) {
     return popup;
   }
+  
+  
+  // Enumerates our configure tree in preparation for searching it
+  private final List<DefaultMutableTreeNode> getSearchNodes(DefaultMutableTreeNode root) {
+    List<DefaultMutableTreeNode> searchNodes = new ArrayList<DefaultMutableTreeNode>();
+
+    Enumeration<?> e = root.preorderEnumeration();
+    while(e.hasMoreElements()) {
+      searchNodes.add((DefaultMutableTreeNode)e.nextElement());
+    }
+    return searchNodes;
+  }
+  
+  
+  // Checks a single string against our search parameters
+  public final boolean checkString(String target, String searchString) {
+    if (matchCase) {
+      return target.contains(searchString);
+    } 
+    else {
+      return target.toLowerCase().contains(searchString.toLowerCase());
+    }         
+  }
+  
+  
+  // Checks a node of the tree against our search parameters
+  public final boolean checkNode(DefaultMutableTreeNode node, String searchString) {
+    Configurable c = null;
+    c = (Configurable) node.getUserObject();
+
+    if (matchNames) {
+      String objectName = c.getConfigureName();
+      if (objectName != null && checkString(objectName, searchString)) {
+        return true;
+      }
+    }
+    
+    if (matchTypes) {
+      String className = getConfigureName(c.getClass());
+      if (className != null && checkString(className, searchString)) {
+        return true;
+      }
+    }
+    
+    return false;
+  }
+  
+  
+  // Search through the tree, starting at the currently selected location (and wrapping around if needed)
+  // Compare nodes until we find our search string (or have searched everything we can search)
+  public final DefaultMutableTreeNode findNode(String searchString) {
+    List<DefaultMutableTreeNode> searchNodes = getSearchNodes((DefaultMutableTreeNode)getModel().getRoot());
+    DefaultMutableTreeNode currentNode = (DefaultMutableTreeNode)getLastSelectedPathComponent();
+
+    DefaultMutableTreeNode foundNode = null;
+    int bookmark = -1;    
+
+    if (currentNode != null) {
+      for (int index = 0; index < searchNodes.size(); index++) {
+        if (searchNodes.get(index) == currentNode) {
+          bookmark = index;
+          break;
+        }
+      }
+    }
+
+    for (int index = bookmark + 1; index < searchNodes.size(); index++) {  
+      if (checkNode(searchNodes.get(index), searchString)) {
+        foundNode = searchNodes.get(index);    
+        break;
+      }
+    }
+
+    if (foundNode == null) {
+      for (int index = 0; index <= bookmark; index++) {    
+        if (checkNode(searchNodes.get(index), searchString)) {
+          foundNode = searchNodes.get(index);    
+          break;
+        }
+      }
+    }
+    
+    return foundNode;
+  }  
+  
+  
+  // Finds out how many total nodes match the search string
+  public final int getNumMatches(String searchString) {    
+    List<DefaultMutableTreeNode> searchNodes = getSearchNodes((DefaultMutableTreeNode)getModel().getRoot());
+    int hits = 0;
```

Comment:
Can be rewritten using Java 8's stream API, the readability is slightly better then:
```
    return (int) searchNodes
      .stream() // entry point of the Stream API
      .filter(node -> checkNode(node, searchString)) // leaves only the elements in the stream that match the predicate
      .count(); // counts elements in the stream, returns a long, that's why it's casted to int a few lines above
```

Reply:
@yanlyub Can you point me in the direction of where I'd get the stream api figured out?

Reply:
This guide looks good but you don't have to get too deep into that topic, the most common methods you will need are `stream()`, `filter()`, `map()`, `count()`, `findFirst()`, `forEach()`, if you roughly understand how they work you can look up the rest later: https://stackify.com/streams-guide-java-8/

Or just learn it as you go, in this particular example we first `filter()` the elements we want (same thing you did with the `if` inside the for loop), then we `count()` what is left (like you did with the `hits++`).

Reply:
Well that certainly made it shorter!